yt_dlp/extractor/zdf.py

```diff
@@ -349,108 +410,160 @@ class ZDFIE(ZDFBaseIE):
+          encryption
+          liveMediaType
+          label
+        }
+        id
+      }
+    }
+  }
+}
+    '''
+
     def _real_extract(self, url):
         video_id = self._match_id(url)
+        video_data = self._download_graphql(video_id, 'video metadata', body={
+            'operationName': 'VideoByCanonical',
+            'query': self._GRAPHQL_QUERY,
+            'variables': {'canonical': video_id},
+        })
+        # TODO: If there are multiple PTMD templates,
+        # usually one of them is a sign-language variant of the video.
+        # The format order works out fine as is and prefers the "normal" video,
+        # but this should probably be made more explicit.
+        ptmd_templates = traverse_obj(
+            video_data,
+            ('data', 'videoByCanonical', 'currentMedia', 'nodes', ..., 'ptmdTemplate'))
+        ptmd_data = self._extract_ptmd(
+            'https://api.zdf.de', ptmd_templates, video_id,
+            self._get_api_token(video_id))
+        # We can't use the ID from PTMD extraction as the video ID
+        # because it is not available during playlist extraction.
+        # We fix it here manually instead of inside the method
+        # because other extractors do rely on using it as their ID.
+        ptmd_data['_old_archive_ids'] = [ptmd_data['id']]
+        del ptmd_data['id']
 
-        webpage = self._download_webpage(url, video_id, fatal=False)
-        if webpage:
-            player = self._extract_player(webpage, url, fatal=False)
-            if player:
-                return self._extract_regular(url, player, video_id, query={'profile': 'player-3'})
+        return {
+            'id': video_id,
+            **ptmd_data,
+            **traverse_obj(video_data, ('data', 'videoByCanonical', {
+                'title': ('title', {str}),
+                'description': (('leadParagraph', ('teaser', 'description')), any, {str}),
+                'timestamp': ('editorialDate', {parse_iso8601}),
+                'thumbnails': ('teaser', 'image', 'list', {self._extract_thumbnails}),
+                'episode_number': ('episodeInfo', 'episodeNumber', {int_or_none}),
+                'season_number': ('episodeInfo', 'seasonNumber', {int_or_none}),
+                'series': ('smartCollection', 'title', {str}),
+                'series_id': ('smartCollection', 'canonical', {str}),
+                'chapters': ('currentMedia', 'nodes', 0, {self._extract_chapters}),
+            })),
+        }
 
-        return self._extract_mobile(video_id)
+    def _extract_chapters(self, data):
+        chapter_marks = traverse_obj(data, ('streamAnchorTags', 'nodes'))
+        chapter_marks.append({'anchorOffset': float_or_none(data.get('duration'))})
+        return [{
+            'start_time': chap.get('anchorOffset'),
+            'end_time': next_chap.get('anchorOffset'),
+            'title': chap.get('anchorLabel'),
+        } for chap, next_chap in zip(chapter_marks, chapter_marks[1:])] or None
```

Comment:
This function is identical to the one in 3Sat?
We can probably make both use the same function from the base class. This was the traversal I used in the 3Sat extractor:
```py
            **traverse_obj(video_target, {
                'duration': ('duration', {int_or_none}),
                'chapters': ('streamAnchorTag', lambda _, v: v['anchorOffset'], {
                    'start_time': ('anchorOffset', {float_or_none}),
                    'title': ('anchorLabel', {str}),
                }),
            }),
```
we don't need to write the end time as yt-dlp core will take care of that.

Reply:
I've unified them into a single function.

> we don't need to write the end time as yt-dlp core will take care of that.

That is very useful. I went by the documentation which made it seem like `end_time` is required. Maybe that's something that can be updated in one of those "Misc" PRs.